### PR TITLE
Avoid redundant signature expansions

### LINKED ISSUE #2287: Redundant keys/signatures in PrecheckVerifier
### Problem

[PrecheckVerifier.hasNecessarySignatures()](https://github.com/hashgraph/hedera-services/blob/b81fb2c009545f59e6c72e32936ee92359c8b7f9/hedera-node/src/main/java/com/hedera/services/sigs/verification/PrecheckVerifier.java#L71) obtains duplicate keys, creates and verifies duplicate signatures for payment queries. The culprit is [PrecheckKeyReqs. getRequiredKeys()](https://github.com/hashgraph/hedera-services/blob/b81fb2c009545f59e6c72e32936ee92359c8b7f9/hedera-node/src/main/java/com/hedera/services/sigs/verification/PrecheckKeyReqs.java#L81) that adds the same key twice to its result.
The issues is easily reproducible with the [HelloWorldSpec](https://github.com/hashgraph/hedera-services/blob/b81fb2c009545f59e6c72e32936ee92359c8b7f9/test-clients/src/main/java/com/hedera/services/bdd/suites/crypto/HelloWorldSpec.java#L43) test.

### Solution

Avoid adding duplicate keys to `getRequiredKeys()` result (`HashSet`?).

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by anighanta**
Skip the signature expansion of other party keys if the party matches the payer

Signed-off-by: anighanta <anirudh.ghanta@hedera.com>

**Description**:

**Related issue(s)**:

Fixes #2287, #2291, #2292 

**Notes for reviewer**:

Currently we repeat the signature expansion of the payer if it is also involved in the transaction other than just paying for it. 
 one Example : 
 ```
cryptoTransfer(
					tinyBarsFromTo(sender, receiver , 1L))
					.payingWith(sender)
``` 
we repeat for `sender` 

Added checks to avoid this.

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by MarcKriguerAtHedera — APPROVED** [review]
In general, this looks good, but please also get someone who is more familiar with the "business logic" to also approve this before merging.

File: hedera-node/src/main/java/com/hedera/services/sigs/verification/PrecheckKeyReqs.java
Comment (OlegMazurov) [inline hedera-node/src/main/java/com/hedera/services/sigs/verification/PrecheckKeyReqs.java:90]: This may not work for `JKey`s that do not provide `equals()`, f.e. `JEd25519Key`

Comment (anighanta) [reply hedera-node/src/main/java/com/hedera/services/sigs/verification/PrecheckKeyReqs.java:90]: We use Object.equals by default which works. Wrote a unit test here hedera-node/src/test/java/com/hedera/services/sigs/verification/PrecheckKeyReqsTest.java -> `usesBothOrderForQueryPayments()`

File: hedera-node/src/main/java/com/hedera/services/legacy/core/jproto/JEd25519Key.java
Comment (OlegMazurov) [inline hedera-node/src/main/java/com/hedera/services/legacy/core/jproto/JEd25519Key.java:86]: `Object.hash()` is redundant. Note that its argument is a boxed integer value.

Comment (anighanta) [reply hedera-node/src/main/java/com/hedera/services/legacy/core/jproto/JEd25519Key.java]: I see that `Objects.hash()` internally does `Arrays.hashCode()`

Comment (OlegMazurov) [reply hedera-node/src/main/java/com/hedera/services/legacy/core/jproto/JEd25519Key.java]: That's right only it's `Arrays.hashCode(Object a[])` and we need `Arrays.hashCode(byte a[])`.

**Review by OlegMazurov — APPROVED** [review]

**Review by MarcKriguerAtHedera — DISMISSED** [review]
Somehow my "Looks Good" review (yesterday) didn't get the approval bit specified.  Setting it now.

File: hedera-node/src/main/java/com/hedera/services/sigs/verification/PrecheckKeyReqs.java
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/sigs/verification/PrecheckKeyReqs.java:90]: Since streams are shockingly expensive, do you think we could, instead, update `addQueryPaymentKeys()` to iterate through the non-payer keys in `otherResult.getOrderedKeys()` and only add a key `nonPayerKey` if `!nonPayerKey.equals(keys.get(0))`?

(This would eliminate the duplication when the `CryptoTransfer` payer and query fee payer are the same; which will be the only duplication 99.99% of the time 😁 ).

Then we could revert this change and avoid the hit from streams.

**Review by tinker-michaelj — COMMENTED** [review]
Overall looks great, very precise! Added one comment to consider avoiding use of Streams API.

File: hedera-node/src/main/java/com/hedera/services/sigs/verification/PrecheckKeyReqs.java
Comment (anighanta) [reply hedera-node/src/main/java/com/hedera/services/sigs/verification/PrecheckKeyReqs.java:90]: internal duplication could happen when say a token create has multiple custom fees and the fee collectors are same for more than one custom fees.

Comment (tinker-michaelj) [reply hedera-node/src/main/java/com/hedera/services/sigs/verification/PrecheckKeyReqs.java:90]: > internal duplication could happen in non-payer keys when say a token create has multiple custom fees and the fee collectors are same for more than one custom fees. But yes this is rare.

A `TokenCreate` txn will never satisfy the `isQueryPayment.test()` tho, right?

Comment (anighanta) [reply hedera-node/src/main/java/com/hedera/services/sigs/verification/PrecheckKeyReqs.java:90]: true

**Review by tinker-michaelj — APPROVED** [review]
LGTM!

**Pull request merged**
